presto-native-execution/presto_cpp/main/types/FunctionMetadata.cpp

```diff
@@ -0,0 +1,222 @@
+    jsonBasedUdfFunctionMetadata.functionKind = functionKind;
+    jsonBasedUdfFunctionMetadata.routineCharacteristics =
+        getRoutineCharacteristics(
+            *functionSignature, functionName, functionKind);
+
+    updateFunctionMetadata(
+        functionName, *functionSignature, jsonBasedUdfFunctionMetadata);
+    jsonBasedUdfFunctionMetadataList.emplace_back(jsonBasedUdfFunctionMetadata);
+  }
+  return jsonBasedUdfFunctionMetadataList;
+}
+
+const std::vector<protocol::JsonBasedUdfFunctionMetadata>
+getWindowFunctionMetadata(
+    const std::string& functionName,
+    const std::vector<FunctionSignaturePtr>& windowFunctionSignatures) {
+  std::vector<protocol::JsonBasedUdfFunctionMetadata>
+      jsonBasedUdfFunctionMetadataList;
+  jsonBasedUdfFunctionMetadataList.reserve(windowFunctionSignatures.size());
+  const protocol::FunctionKind functionKind = protocol::FunctionKind::WINDOW;
+
+  for (const auto& windowFunctionSignature : windowFunctionSignatures) {
+    protocol::JsonBasedUdfFunctionMetadata jsonBasedUdfFunctionMetadata;
+    jsonBasedUdfFunctionMetadata.functionKind = functionKind;
+    jsonBasedUdfFunctionMetadata.routineCharacteristics =
+        getRoutineCharacteristics(
+            *windowFunctionSignature, functionName, functionKind);
+
+    updateFunctionMetadata(
+        functionName, *windowFunctionSignature, jsonBasedUdfFunctionMetadata);
+    jsonBasedUdfFunctionMetadataList.emplace_back(jsonBasedUdfFunctionMetadata);
+  }
+  return jsonBasedUdfFunctionMetadataList;
+}
+
+const std::vector<protocol::JsonBasedUdfFunctionMetadata> getFunctionMetadata(
+    const std::string& functionName) {
+  if (auto aggregateFunctionSignatures =
+          getAggregateFunctionSignatures(functionName)) {
+    return getAggregateFunctionMetadata(
+        functionName, aggregateFunctionSignatures.value());
+  } else if (
+      auto windowFunctionSignatures =
+          getWindowFunctionSignatures(functionName)) {
+    return getWindowFunctionMetadata(
+        functionName, windowFunctionSignatures.value());
+  } else {
+    auto functionSignatures = getFunctionSignatures();
+    if (functionSignatures.find(functionName) != functionSignatures.end()) {
+      return getScalarFunctionMetadata(
+          functionName, functionSignatures.at(functionName));
+    }
+  }
+  VELOX_UNREACHABLE("Function kind cannot be determined");
+}
+
+} // namespace
+
+void getJsonMetadataForFunction(
+    const std::string& registeredFunctionName,
+    nlohmann::json& jsonMetadataList) {
+  auto functionMetadataList = getFunctionMetadata(registeredFunctionName);
+  auto functionName = getFunctionName(registeredFunctionName);
+  for (const auto& functionMetadata : functionMetadataList) {
+    jsonMetadataList[functionName].emplace_back(functionMetadata);
+  }
+}
+
+json getJsonFunctionMetadata() {
+  auto registeredFunctionNames = getSortedAggregateNames();
+  auto scalarFunctionNames = getSortedScalarNames();
```

Comment:
Just curious : Is it necessary to have a sorted list here ?

Reply:
No it isn't necessary, but the functions return sorted lists already.